### PR TITLE
Add unit tests for CreateAccount and RecoverAccount methods #510

### LINKED ISSUE #510: Add unit tests for CreateAccount and RecoverAccount methods
# Problem

`./geth/account` package has a poor unit tests coverage which should be improved.

# Implementation

There is only one struct that needs to be tested: `Manager` in `geth/account/accounts.go`. This struct depends on `common.NodeManager` and we'd like to get rid of this dependency, thus, please implement unit tests **only** for methods that do not depend on `common.NodeManager`. Identified methods so far: `CreateAccount` and `RecoverAccount`. 

Unit tests should be put in `geth/account/accounts_test.go` file.

# Acceptance Criteria

1. There are no race conditions,
1. Test coverage for identified methods should be decent (as a reference `VerifyAccountPassword` has 89.7% coverage; something over 80% should be fine but follow common sense :)). 

# Notes

Please follow code styles guidelines: [Effective Go](https://golang.org/doc/effective_go.html) and [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments).

To get a coverage for functions, run:
```
$ go test -coverprofile=coverage.out  ./geth/account/...
$ go tool cover -func=coverage.out
```

### PR DESCRIPTION
**Pull request by canercidam**
Unit tests for account creation and recovery using `common.MockNodeManager`

Important changes:
- [x] No race conditions.
- [x] Coverage for `Manager.CreateAccount` and `Manager.RecoverAccount` is over 80%.
- [x] ~~All functions which are independent from `common.NodeManager` are included. (?)~~
- [x] Overall test coverage is over 80%.

Closes #510

### CODE REVIEW

File: geth/account/accounts_test.go [hunk 1]
```diff
@@ -109,3 +111,46 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+
+func TestCreateAndRecoverAccount(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	nodeManager := common.NewMockNodeManager(ctrl)
+
+	// Create an arbitrary keystore for the process
+	keyStoreDir, err := ioutil.TempDir(os.TempDir(), "accounts")
+	require.NoError(t, err)
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+	keyStore := keystore.NewKeyStore(keyStoreDir, 4, 5)
```
Comment (adambabik) [inline geth/account/accounts_test.go]: `4` and `5` look like magic numbers here. `github.com/ethereum/go-ethereum/accounts/keystore` package has `LightScryptN` and `LightScryptP` constants exported which should be fine here.

File: geth/account/accounts_test.go [hunk 2]
```diff
@@ -109,3 +111,46 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+
+	manager := account.NewManager(nodeManager)
```
Comment (adambabik) [inline geth/account/accounts_test.go]: Maybe we can call it `accountManager` or `accManager`? It's confusing with `nodeManager`. It may be just me because `nodeManager` is usually the main `manager` ;)

File: geth/account/accounts_test.go [hunk 3]
```diff
@@ -109,3 +111,46 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	errString := "Non-nil error string"
+	nodeManager.EXPECT().AccountKeyStore().Return(nil, errors.New(errString))
+	_, _, _, err = manager.CreateAccount(password)
+	require.Equal(t, err.Error(), errString)
+
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	addr1, pubKey1, mnemonic, err := manager.CreateAccount(password)
+	require.NoError(t, err)
+
+	// Expect non-nil return values
```
Comment (adambabik) [inline geth/account/accounts_test.go]: This comment is not needed.

File: geth/account/accounts_test.go [hunk 4]
```diff
@@ -109,3 +111,46 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	require.NotNil(t, addr1)
+	require.NotNil(t, pubKey1)
+	require.NotNil(t, mnemonic)
+
+	// Now recover the account using the mnemonic seed and the password
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	addr2, pubKey2, err := manager.RecoverAccount(password, mnemonic)
+	require.NoError(t, err)
+
+	// Expect the same return values
```
Comment (adambabik) [inline geth/account/accounts_test.go]: And this as well as `require.Equal` exactly tells us what we're gonna do.

File: geth/account/accounts_test.go [hunk 5]
```diff
@@ -109,3 +111,46 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	// Create an arbitrary keystore for the process
+	keyStoreDir, err := ioutil.TempDir(os.TempDir(), "accounts")
+	require.NoError(t, err)
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+	keyStore := keystore.NewKeyStore(keyStoreDir, 4, 5)
+
+	manager := account.NewManager(nodeManager)
+	password := "some-pass"
+
+	// Fail if account keystore can't be acquired
```
Comment (adambabik) [inline geth/account/accounts_test.go]: What do you think about extracting failing scenarios to a separate test? Something like `TestCreateAccountFailure()` that could verify this error?

File: geth/account/accounts_test.go [hunk 6]
```diff
@@ -109,3 +111,46 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	nodeManager := common.NewMockNodeManager(ctrl)
+
+	// Create an arbitrary keystore for the process
+	keyStoreDir, err := ioutil.TempDir(os.TempDir(), "accounts")
+	require.NoError(t, err)
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+	keyStore := keystore.NewKeyStore(keyStoreDir, 4, 5)
+
+	manager := account.NewManager(nodeManager)
+	password := "some-pass"
```
Comment (adambabik) [inline geth/account/accounts_test.go]: I would also verify empty string as a password. I think it may be ok but let's make sure as it's a special case. You can also think about extracting various password cases to separate `TestCreateAccount()` function.

**Review by adambabik — CHANGES_REQUESTED** [review]
Nice solution using `common.NewMockNodeManager()`. Please take a look at my comments.

File: geth/account/accounts_test.go (hunk 6 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: I actually did this one but removed later :)
Using an empty string doesn't make a difference in the result, it throws no error and returns a valid result. Shall we also expect it to not fail with an empty password in tests? Or slightly extend the scope of this PR and modify the function to fail?

File: geth/account/accounts_test.go (hunk 5 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Ah yes, makes more sense. Thanks.

File: geth/account/accounts_test.go (hunk 6 above)
Comment (adambabik) [reply geth/account/accounts_test.go]: > Shall we also expect it to not fail with an empty password in tests?

No, this is good I think, at least for now. My point is to just make a unit test for this scenario so that if it changes, we're aware as this is an edge case.

File: geth/account/accounts_test.go [hunk 7]
```diff
@@ -109,3 +111,78 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
 	_, err = acctManager.VerifyAccountPassword(keyStoreDir, address.Hex(), TestConfig.Account3.Password)
 	require.NoError(t, err)
 }
+
+// newTestAccManager returns the account manager and the mock node manager
+// that's being used within it for testing purposes.
+func newTestAccManager(t *testing.T) (accManager *account.Manager, nodeManager *common.MockNodeManager) {
```
Comment (adambabik) [inline geth/account/accounts_test.go]: If it returns multiple managers then maybe `newTestManager`?

File: geth/account/accounts_test.go [hunk 8]
```diff
@@ -109,3 +111,78 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+
+	keyStore, keyStoreDir := newTestKeyStore(t, "accounts")
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+
+	// Don't fail on empty password
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	_, _, _, err := accManager.CreateAccount(password)
+	require.NoError(t, err)
+
+	password = "some-pass"
```
Comment (adambabik) [inline geth/account/accounts_test.go]: For multiple test cases, I recommend using table-driven tests. You can read more about this approach here: https://blog.golang.org/subtests.

Here we have only two cases so it's not a big deal.

**Review by adambabik — APPROVED** [review]
LGTM! Please check out my minor comments. 

Also, a general note is that in some places code comments or some asserts are separated by a blank line which is not necessary. They may be artifacts while moving the code around :)

File: geth/account/accounts_test.go (hunk 8 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Nice tip, thanks! I've actually seen that practice in `TestVerifyAccountPassword`.

However, I had to pair a successful creation and a recovery in `TestCreateAndRecoverAccountSuccess` since recovery uses both the input and one of the outputs of creation. And `TestCreateAndRecoverFail_KeyStore` is different from the first one regarding that. So it doesn't seem useful in this case I suppose.

**Comment by canercidam** [comment]
Adding table-driven `TestSelectAccount`.

**Comment by canercidam** [comment]
> @adambabik:
> ...
> This also makes it ok to write unit tests for other exported methods like SelectAccount. However, as it was not in the original scope, it's up to you :)

I can keep on increasing test coverage if everything looks fine so far. Are we worried about the amount of lines in test file, though? I can also try to separate them into several files.

File: geth/account/accounts_test.go [hunk 9]
```diff
@@ -109,3 +112,154 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	password := "some-pass"
+
+	keyStore, keyStoreDir := newTestKeyStore(t, "accounts")
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	addr, _, _, err := accManager.CreateAccount(password)
+	require.NoError(t, err)
+
+	w := whisper.New(nil)
```
Comment (adambabik) [inline geth/account/accounts_test.go]: In this context, `w` may be hard to figure out what it refers to. I would recommend calling it `shh` which is usually used for Whisper service.

File: geth/account/accounts_test.go [hunk 10]
```diff
@@ -109,3 +112,154 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	// Don't fail on empty password
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	_, _, _, err := accManager.CreateAccount(password)
+	require.NoError(t, err)
+
+	password = "some-pass"
+
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	addr1, pubKey1, mnemonic, err := accManager.CreateAccount(password)
+	require.NoError(t, err)
```
Comment (adambabik) [inline geth/account/accounts_test.go]: No need for newline below.

File: geth/account/accounts_test.go [hunk 11]
```diff
@@ -109,3 +112,154 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+
+	require.NotNil(t, addr1)
+	require.NotNil(t, pubKey1)
+	require.NotNil(t, mnemonic)
+
+	// Now recover the account using the mnemonic seed and the password
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	addr2, pubKey2, err := accManager.RecoverAccount(password, mnemonic)
+	require.NoError(t, err)
```
Comment (adambabik) [inline geth/account/accounts_test.go]: Below this one as well. Please take a look at other places.

**Review by adambabik — APPROVED** [review]

**Comment by adambabik** [comment]
> I can keep on increasing test coverage if everything looks fine so far. Are we worried about the amount of lines in test file, though? I can also try to separate them into several files.

I don't think so. It's ok to put all related tests to `accounts.go` in `accounts_test.go`.

**Comment by canercidam** [comment]
Major changes in the test file since last comments. By the way, I also read all the discussion from #445 to #510 and I'm actually wondering if using a mock like this still conflicts with the idea of unit testing expected to be done here. Without using the mock, there doesn't seem to be a more flexible way of covering error cases sourcing from `common.NodeManager` in tests.

File: geth/account/accounts_test.go [hunk 12]
```diff
@@ -103,9 +107,381 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+// development so this is a good workaround to use with EXPECT().AnyTimes()
+func (s *ManagerTestSuite) reinitMock() {
+	s.nodeManager = newMockNodeManager(s.T())
+	s.accManager.nodeManager = s.nodeManager
+}
+
+func (s *ManagerTestSuite) TestCreateAndRecoverAccount() {
+	s.reinitMock()
+
+	// Don't fail on empty password
```
Comment (adambabik) [inline geth/account/accounts_test.go]: How is this connected to `TestCreateAndRecoverAccount`? I am not sure this belongs here...

File: geth/account/accounts_test.go [hunk 13]
```diff
@@ -103,9 +107,381 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	accManager := NewManager(nodeManager)
+
+	keyStoreDir, err := ioutil.TempDir(os.TempDir(), "accounts")
+	require.NoError(t, err)
+	keyStore := keystore.NewKeyStore(keyStoreDir, keystore.LightScryptN, keystore.LightScryptP)
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+
+	testPassword := "test-password"
+
+	// Initial test - create test account
```
Comment (adambabik) [inline geth/account/accounts_test.go:134]: In the previous version, I believe it was in a separate test case together with an empty password case. I think it was a better solution.

File: geth/account/accounts_test.go [hunk 14]
```diff
@@ -103,9 +107,381 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+func (s *ManagerTestSuite) TestSelectAccount() {
+	s.reinitMock()
+
+	testCases := []struct {
+		name                  string
+		accountKeyStoreReturn []interface{}
+		whisperServiceReturn  []interface{}
+		address               string
+		password              string
+		fail                  bool
```
Comment (adambabik) [inline geth/account/accounts_test.go]: Alternatively, you can replace `fail bool` with `err error` and check if the returned error is equal. In case of a `nil`, you can check there is no error.

File: geth/account/accounts_test.go [hunk 15]
```diff
@@ -103,9 +107,381 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	s.T().Run("success", func(t *testing.T) {
+		acc, err := s.accManager.SelectedAccount()
+		s.NoError(err)
+		s.NotNil(acc)
+
+		err = s.accManager.ReSelectAccount()
+		s.NoError(err)
+	})
+
+	s.T().Run("ReSelect_fail", func(t *testing.T) {
```
Comment (adambabik) [inline geth/account/accounts_test.go]: As this is specific to Whisper service fail, maybe `ReSelect_fail_whisper`?

File: geth/account/accounts_test.go [hunk 16]
```diff
@@ -103,9 +107,381 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+			} else {
+				s.NoError(err)
+				s.NotNil(childAddr)
+				s.NotNil(childPubKey)
+			}
+		})
+	}
+}
+
+func (s *ManagerTestSuite) TestSelectedAndReSelectAccount() {
```
Comment (adambabik) [inline geth/account/accounts_test.go:350]: I would split `SelectAccount` and `ReselectAccount` into two tests. It's a little bit hard here to figure our all scenarios being tested. Maybe you can add comments to some of them? E.g. `// test reselect when there is no account selected` etc.

File: geth/account/accounts_test.go [hunk 17]
```diff
@@ -103,9 +107,381 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+			"wrong-password",
+			true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		s.T().Run(testCase.name, func(t *testing.T) {
+			s.reinitMock()
+			s.nodeManager.EXPECT().AccountKeyStore().Return(testCase.accountKeyStoreReturn...).AnyTimes()
+			acc, keyStore, err := s.accManager.AddressToDecryptedAccount(testCase.address, testCase.password)
```
Comment (adambabik) [inline geth/account/accounts_test.go]: We should be able to verify `acc` value.

**Review by adambabik — CHANGES_REQUESTED** [review]

**Comment by adambabik** [comment]
@canercidam I think it's ok in this case. Overall, you only mock `NodeManager` and the rest is regular objects so many of these tests are integration tests on the lowest level possible + unit tests.

File: geth/account/accounts_test.go (hunk 13 above)
Comment (canercidam) [reply geth/account/accounts_test.go:134]: I moved the account creation test to `TestManagerTestSuite` in order to create an account while setting up the test. Otherwise, one has to create an account in the beginning of almost every test function. And I also didn't want it to rely on a predefined account and password but rather have it created dynamically in the beginning of every test execution.

File: geth/account/accounts_test.go (hunk 12 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Since the account is created in `TestManagerTestSuite`, this one is just testing two more cases for account creation. And it also tests recovering of the previously created account plus a failing case.

File: geth/account/accounts_test.go (hunk 14 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Makes sense to me. I'll do a rewriting for all regarding this.

File: geth/account/accounts_test.go (hunk 16 above)
Comment (canercidam) [reply geth/account/accounts_test.go:350]: I'll make it more clear.

File: geth/account/accounts_test.go (hunk 13 above)
Comment (canercidam) [reply geth/account/accounts_test.go:134]: I can alternatively rely on `status-go/testing.TestConfig` as in `TestVerifyAccountPassword` if that sounds better.

File: geth/account/accounts_test.go (hunk 17 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: This looks like the best way to do it but please confirm (there's a typo - `keyStore` should be `key`):
```go
s.Equal(acc.Address, key.Address)
s.keyStore.Find(acc)
```

File: geth/account/accounts_test.go (hunk 13 above)
Comment (adambabik) [reply geth/account/accounts_test.go:134]: Yeah, I got your point. It's ok to create an account that will be used over and over and store it in `ManagerTestSuite`. So, we can leave it here as it is. Let's continue below on `TestCreateAndRecoverAccount` comments.

File: geth/account/accounts_test.go (hunk 12 above)
Comment (adambabik) [reply geth/account/accounts_test.go]: The thing is that this test case has nothing to do with recovering an account. Maybe let's split it into `TestCreateAccount` and `TestRecoverAccount`? 

In `TestCreateAccount` we can have an empty password case and failed `CreateAccount()` call:
```go
s.nodeManager.EXPECT().AccountKeyStore().Return(nil, testErrKeyStore)
_, _, _, err = s.accManager.CreateAccount(s.password)
s.Equal(err, testErrKeyStore)
```

In `TestRecoverAccount`, we will just test recovering an account, so only `accManager.RecoverAccount()` calls. Does it sound ok?

File: geth/account/accounts_test.go (hunk 17 above)
Comment (adambabik) [reply geth/account/accounts_test.go]: `s.Equal(acc.Address, key.Address)` check should be ok.

Regarding, `s.keyStore.Find(acc)`, per `AddressToDecryptedAccount` comment, it requires the key file for a given address to be prior to node start. So, I think it's not necessary as this call should fail if this check is invalid anyway.

File: geth/account/accounts_test.go (hunk 12 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Surely, splitting.

**Comment by canercidam** [comment]
@adambabik ping

**Review by adambabik — APPROVED** [review]
Great stuff! I really like the comments and the way you structured code blocks. It's very easy to read and understand.

**Comment by adambabik** [comment]
@canercidam great timing :)

**Comment by canercidam** [comment]
@adambabik Ah, just at the same minute. 😄 Cheers! Thanks for taking the time to review.

File: geth/account/accounts_test.go [hunk 18]
```diff
@@ -103,9 +107,386 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	s.Equal(s.address, addr)
+	s.Equal(s.pubKey, pubKey)
+
+	s.nodeManager.EXPECT().AccountKeyStore().Return(nil, testErrKeyStore)
+	_, _, err = s.accManager.RecoverAccount(s.password, s.mnemonic)
+	s.Equal(testErrKeyStore, err)
+}
+
+func (s *ManagerTestSuite) TestSelectAccount() {
+	s.reinitMock()
```
Comment (JekaMas) [inline geth/account/accounts_test.go]: Shell we extract `s.reinitMock()` into `SetupTest`?

File: geth/account/accounts_test.go [hunk 19]
```diff
@@ -103,9 +107,386 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	for _, testCase := range testCases {
+		s.T().Run(testCase.name, func(t *testing.T) {
+			s.reinitMock()
+			s.nodeManager.EXPECT().AccountKeyStore().Return(testCase.accountKeyStoreReturn...).AnyTimes()
+			s.nodeManager.EXPECT().WhisperService().Return(testCase.whisperServiceReturn...).AnyTimes()
+			err := s.accManager.SelectAccount(testCase.address, testCase.password)
+			if testCase.expectedError != nil {
+				s.Equal(testCase.expectedError, err)
+			} else {
+				s.NoError(err)
```
Comment (JekaMas) [inline geth/account/accounts_test.go]: It looks like duplicate check.

File: geth/account/accounts_test.go [hunk 20]
```diff
@@ -103,9 +107,386 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	s.nodeManager.EXPECT().WhisperService().Return(s.shh, nil)
+	err := s.accManager.Logout()
+	s.NoError(err)
+
+	s.nodeManager.EXPECT().WhisperService().Return(nil, testErrWhisper)
+	err = s.accManager.Logout()
+	s.Equal(testErrWhisper, err)
+}
+
+func (s *ManagerTestSuite) TestAccounts() {
```
Comment (JekaMas) [inline geth/account/accounts_test.go]: `TestAccounts` actually doesn't say much about the test. Could you rename it accoording to test case below?

**Review by JekaMas — COMMENTED** [review]
All changes are minor. If you're ready to make them i'll wait. Otherwice 'LGTM'.

File: geth/account/accounts_test.go (hunk 20 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Every test function is named using the common naming convention such as `TestFunctionName` and function name is only `Accounts` in this case. I hope it makes sense.

File: geth/account/accounts_test.go (hunk 19 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Ah yes, it looks possible to reduce in this case. Changing.

File: geth/account/accounts_test.go (hunk 18 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Extracting? It's a separate function to use where necessary in order to drop all calls in `gomock.Controller`. Comment lines above `reinitMock` should make sense.

Comment (themue) [reply geth/account/accounts_test.go]: Moving it into a function `SetupTest()` which automatically is executed for each test makes the manual calling in the beginning of each test obsolete. So not a direct extracting, but moving into a different method. That's what he meant.

File: geth/account/accounts_test.go (hunk 20 above)
Comment (themue) [reply geth/account/accounts_test.go]: Hehe, yeah, typical trouble due to the naming convention. That's why I personally for my code also like little comments for each test.

But here for me it's okay.

**Review by themue — CHANGES_REQUESTED** [review]
Only one change, otherwise LGTM.

File: geth/account/accounts_test.go (hunk 20 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Hmm. Let's just leave a short comment to make it more obvious.

File: geth/account/accounts_test.go (hunk 18 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Alright then. That surely sounds like a better way to do it.

File: geth/account/accounts_test.go [hunk 21]
```diff
@@ -103,9 +107,373 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	address  string
+	pubKey   string
+	mnemonic string
+}
+
+// reinitMock is for reassigning a new mock node manager to account manager.
+// Stating the amount of times for mock calls kills the flexibility for
+// development so this is a good workaround to use with EXPECT().Func().AnyTimes()
+func (s *ManagerTestSuite) reinitMock() {
+	s.nodeManager = newMockNodeManager(s.T())
```
Comment (themue) [inline geth/account/accounts_test.go:185]: Only twp lines. So move into `SetupTest()` (including comment).

File: geth/account/accounts_test.go [hunk 22]
```diff
@@ -103,9 +107,373 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
...
+	s.nodeManager.EXPECT().WhisperService().Return(s.shh, nil)
+	err := s.accManager.Logout()
+	s.NoError(err)
+
+	s.nodeManager.EXPECT().WhisperService().Return(nil, testErrWhisper)
+	err = s.accManager.Logout()
+	s.Equal(testErrWhisper, err)
+}
+
+// TestAccounts tests cases for (*Manager).Accounts
```
Comment (themue) [inline geth/account/accounts_test.go]: Top 👍 . Only interpunction missing. Yeah, I know, sometimes pedantic. 😁

**Review by themue — APPROVED** [review]
Only little final change, but overall LGTM.

File: geth/account/accounts_test.go (hunk 21 above)
Comment (canercidam) [reply geth/account/accounts_test.go:185]: I mean, `reinitMock` is also called multiple times in test functions. It's mainly for dropping all calls whenever needed and use the mock as an ever-changing stub for different cases in a test function, especially during table-driven tests. That's why I objected to "extracting" :)

Since it's also called in the beginning of every test function, it sounded reasonable to use `SetupTest`, though.

Comment (themue) [reply geth/account/accounts_test.go:185]: Ah, okay, missed it. So it's fine.

File: geth/account/accounts_test.go (hunk 22 above)
Comment (canercidam) [reply geth/account/accounts_test.go]: Do you mean like `Manager.Accounts` instead?

Comment (themue) [reply geth/account/accounts_test.go]: Oh, eh, no. Just

```
// TestAccounts tests cases for (*Manager).Accounts.
```

I do like full sentences for comments, also no friend `// do this and that` (w/o interpunction and starting with a lower-case letter in case of an inline comment). As I said, I'm personally a bit pedantic. 🤣

Comment (canercidam) [reply geth/account/accounts_test.go]: Ah, okay 😆 Adding the dot, no worries. lol

**Review by JekaMas — APPROVED** [review]
